### PR TITLE
Add support for sparse metadata labelling layers

### PR DESCRIPTION
**Pull request by lmcinnes**
A simple approach to labelling when dealing with sparse metadata (e.g. bag-of-words matrices). The approach is similar to BERTopic's c-TF-IDF but uses the more advanced options available from the semi-supervised ``InformationWeightTransformer`` in ``vectorizers``, and (since we do not assume joint embeddings) doesn't look at closeness of selected keywords in the joint space.

### CODE REVIEW

File: thisnotthat/map_cluster_labelling.py [hunk 1]
```diff
@@ -803,6 +807,36 @@ def text_labels_from_per_sample_labels(
     return labels
 
 
+def text_labels_from_sparse_metadata(
+    pointset_layers: List[List[npt.NDArray]],
```
Comment (hamelin) [inline thisnotthat/map_cluster_labelling.py:811]: Why make these three parameters strictly positional?

File: thisnotthat/map_cluster_labelling.py [hunk 2]
```diff
@@ -1380,3 +1414,187 @@ def labels_for_display(self):
...
+    location_layers: List of Arrays of shape (n_cluster_in_layer, n_map_features)
+        A list of layers; each layer is an array of locations in the map representation to place the labels of that layer
+
+    labels_for_display: List of list of labels
+        A list of layers; each layer is a list of labels; each label is formatted for display use by ``label_formatter``
+    """
+
+    def __init__(
+        self,
+        source_vectors: npt.ArrayLike,
```
Comment (hamelin) [inline thisnotthat/map_cluster_labelling.py:1520]: Again: why strictly positional parameters? With such a long list of formal parameters, I would certainly use as many named parameters as I could...

File: thisnotthat/map_cluster_labelling.py [hunk 3]
```diff
@@ -1380,3 +1414,187 @@ def labels_for_display(self):
...
+            map_representation,
+            cluster_map_representation=cluster_map_representation,
+            umap_metric=vector_metric,
+            umap_n_components=umap_n_components,
+            umap_n_neighbors=umap_n_neighbors,
+            hdbscan_min_samples=hdbscan_min_samples,
+            hdbscan_min_cluster_size=hdbscan_min_cluster_size,
+            random_state=random_state,
+        )
+        (
```
Comment (hamelin) [inline thisnotthat/map_cluster_labelling.py:1563]:
```suggestion

        (
```

Comment (hamelin) [reply thisnotthat/map_cluster_labelling.py:1563]: Reading this code I thought at first these parentheses were for calling a function returned by `build_fine_grained_cluster_centers` (which would be some wild-west idiom, but I can get behind it), and it took me a while to see it was to accept three return values from `build_cluster_layers`. I believe this line skip facilitates reading.

**Review by hamelin — CHANGES_REQUESTED** [review]
The notebook file `doc/sparse_metadata_cluster_labels.ipynb` is corrupted: only a prefix of the text seems to have been included by either Git or Github.

**Review by hamelin — APPROVED** [review]
I'll approve to avoid putting myself in the way, but I have a few questions dispersed between the code lines that I believe should be addressed.

File: thisnotthat/map_cluster_labelling.py [hunk 4]
```diff
@@ -1380,3 +1414,187 @@ def labels_for_display(self):
...
+    fetaure_name_dictionary: dict
+        A dictionary mapping column indices of the sparse matrix to feature names. For example, if the sparse matrix
+        were the output of sklearn's ``CountVectorizer`` the dict would be
+        ``{idx: word for word, idx in model.vocabulary_.items()}``.
+
+    vector_metric: str (optional, default = "cosine")
+        The metric to use on the source vector space.
+
+    cluster_map_representation: bool (optional, default = False)
+        Whether to directly cluster the map representation, or use UMAP to generate a representation for clustering
```
Comment (jc-healy) [inline thisnotthat/map_cluster_labelling.py:1447]:
```suggestion
If set to True, cluster your raw map representation.  If set to False apply UMAP first to generate a representation for clustering.
```
It's less eloquent and equivalent but a touch clearer (in my head at least) as I had to read this twice.

**Review by jc-healy — APPROVED** [review]
Thanks for writing this up.  It will be handy to have in the toolbox.

**Pull request merged**